leader/leader.go

```diff
@@ -1,59 +1,88 @@
 package leader
 
 import (
-	"encoding/json"
-	"io/ioutil"
-	"net/http"
+	"context"
+	"fmt"
 	"os"
-)
+	"sync/atomic"
+	"time"
 
-const defaultURL = "http://localhost:4040/"
+	"github.com/cenkalti/backoff/v4"
+	"github.com/hashicorp/go-hclog"
+	operator_leader "github.com/operator-framework/operator-lib/leader"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+)
 
-type LeaderElector struct {
-	URL string
+type Elector interface {
+	// IsLeader returns whether this host is the leader
+	IsLeader() (bool, error)
 }
 
-type LeaderResponse struct {
-	Name string `json:"name"`
+type LeaderForLife struct {
+	isLeader atomic.Value
 }
 
-// New returns a LeaderElector with the default service endpoint
-func New() *LeaderElector {
-	return &LeaderElector{
-		URL: defaultURL,
-	}
-}
+// New returns a Elector that uses the operator-sdk's leader for life elector
+func New(ctx context.Context, logger hclog.Logger, clientset kubernetes.Interface) *LeaderForLife {
+	le := &LeaderForLife{}
+	le.isLeader.Store(false)
 
-// NewWithURL returns a LeaderElector with a custom service endpoint URL
-func NewWithURL(URL string) *LeaderElector {
-	return &LeaderElector{
-		URL: URL,
+	if err := setLeaderLabel(ctx, clientset, false); err != nil {
+		logger.Warn("trouble setting leader=false", "err", err)
 	}
+	go func() {
+		// This function blocks until this replica becomes the "leader", which
+		// means it creates a ConfigMap with an ownerref. Another replica can
+		// become the leader when the current leader replica stops running, and
+		// the Kubernetes garbage collector deletes the vault-k8s-leader
+		// ConfigMap.
+		err := operator_leader.Become(ctx, "vault-k8s-leader")
+		if err != nil {
+			logger.Error("trouble becoming leader:", "error", err)
+			return
+		}
+		le.isLeader.Store(true)
+		if err := setLeaderLabel(ctx, clientset, true); err != nil {
+			logger.Warn("trouble setting leader=true label", "err", err)
+		}
+	}()
+
+	return le
 }
 
 // IsLeader returns whether this host is the leader
-func (le *LeaderElector) IsLeader() (bool, error) {
-	resp, err := http.Get(le.URL)
-	if err != nil {
-		return false, err
-	}
-	defer resp.Body.Close()
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return false, err
-	}
-	leaderResp := &LeaderResponse{}
-	err = json.Unmarshal(body, leaderResp)
-	if err != nil {
-		return false, err
-	}
-	hostname, err := os.Hostname()
-	if err != nil {
-		return false, err
+func (le *LeaderForLife) IsLeader() (bool, error) {
+	leaderStatus := le.isLeader.Load().(bool)
+	return leaderStatus, nil
+}
+
+func setLeaderLabel(ctx context.Context, clientset kubernetes.Interface, label bool) error {
+	namespace := os.Getenv("NAMESPACE")
+	if namespace == "" {
```

Comment:
It looks like this comment was on an old commit, but as for checking for POD_NAME, Become() will return an error if POD_NAME isn't set. And since we're handling that error with a finite amount of retries, it'll eventually case the injector to exit non-zero.